### ISSUE TITLE
Un-narrowing optional type when defining nested function within a conditional

### ISSUE THREAD

**Issue by mishok13** [python/mypy#15251]
**Bug Report**

When defining a nested function under a branch that accesses argument passed to outer function, the type narrowing resets to the original type.

**To Reproduce**

[See the playground](https://mypy-play.net/?mypy=1.2.0&python=3.11&flags=strict). Code posted here as well.

```python
from functools import wraps
from typing import Any, Callable

def d(f: Callable[..., Any] | None) -> None:
    if f is not None: # <-- narrowed to Callable
        def inner() -> Any:
            return f() # <-- resolved as Callable | None
    return None
```

**Expected Behavior**

Inner function retains narrowed type.

**Actual Behavior**

Function `f`'s type is resolved as `Callable | None` leading to this error:

```
main.py:9: error: "None" not callable  [misc]
```

**Your Environment**

- Mypy version used: 1.2, 1.3
- Mypy command-line flags: `--strict`
- Mypy configuration options from `mypy.ini` (and other config files): no additional config
- Python version used: 3.11, 3.10

**Comment by AlexWaygood** [comment]
Duplicate of #2608

**Comment by AlexWaygood** [comment]
(Your code type checks fine on mypy master. The fix for #2608 will be available in the next release of mypy :)
